var/spack/repos/builtin/packages/fenics-dolfinx/package.py

```diff
@@ -21,14 +21,15 @@ class FenicsDolfinx(CMakePackage):
     variant("parmetis", default=False, description="parmetis support")
     variant("slepc", default=False, description="slepc support")
 
-    depends_on("cmake@3.12:")
+    depends_on("cmake@3.18:")
```

Comment:
```suggestion
    depends_on("cmake@3.18:", type='build')
```

Reply:
We've found using `type='build'` with CMake problematic. The issue is that if a user builds against Basix but with a different CMake bad things can happen because different CMake versions can have different dependency detection (different FindFoo.cmake), which then messes up with`fooTargets.cmake` up breaks builds.

Reply:
Hmm, how does that relate with the type being `build` instead of `build, link`? Currently we unify versions of build dependencies.

Reply:
If the type is `build, link`, then `spack load fenics-basix` loads CMake. If the type is `build`, then CMake doesn't get loaded when `fenics-basix` is loaded and there is a good chance that a different version of CMake is in the system path. This can cause failures when an older system CMake doesn't support the targets that are in the generated ` BasixTargets.cmake` file.

Reply:
Can you go through your workflow? I'm not entirely sure I understand how `spack load` enters into the picture. CMake minimum versions should be declared in `CMakeLists.txt`, right?

Reply:
Yes, the minimum version could (and perhaps should) be added to `CMakeLists.txt` for programs that link to `fenics-basix`, etc. However, if a program `foo` links to `fenics-basix`, I don't know how to check that the CMakeLists.txt file for `foo` enforces a new enough version of CMake in order to parse `BasixTargets.cmake` correctly (which is generated by CMake when configuring `fenics-basix`).

The issue came up with `fenics-dolfinx`, which has more dependencies, but I'll sketch the issue.
1. Library `foo` has a `CMakeLists.txt` that contains `find_package(bar); target_link_libraries(foo  PUBLIC bar::bar)` 
2. When configuring and installing `foo`, CMake (using the CMake version specified in the Spack package ) creates and installs a file `fooConfig.cmake` 
   ```
   set_target_properties(food::foo PROPERTIES
    INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
    INTERFACE_LINK_LIBRARIES "bar::bar")
   ```
3. `spack load foo` brings `foo` into the library path
4. User attempts to build a program against `foo` with the (old) systems CMake, which has a `FindBar.cmake` file, but it is too old to provide the `bar::bar` targets. CMake can't resolve `bar::bar` and build fails.

I'm happy to switch back to `type=build` if that will help move this PR along, and document somewhere for users that they must use some minimum CMake version. We did see users frequently hitting build failures due to mis-matched CMake versions with `fenics-dolfinx`. The trigger was the Spack HDF5 package recently switching from autotools to CMake builds, which now generates a `hdf5-config.cmake` and which was handled differently by different CMake versions.

Reply:
> I'm happy to switch back to type=build if that will help move this PR along, and document somewhere for users that they must use some minimum CMake version.

Yeah, let's do that. Using `spack load` has little to do with the installation logic, i.e. if I understand correctly you are talking about a workflow where users try to consume manually some software that has been installed by Spack. This software _can_ be installed with CMake version X, but if version Y > X gets used this imposed an implicit constraint on dependents that need to use version Y or greater too. 

In general, modeling the dependency with a different type than it should be to get something in the PATH has repercussion on other legitimate use cases. For instance this package will deploy CMake when installed from a buildcache, which is unexpected.

Also, you might want to check `spack develop` which should take care of these issues for you. In case the [tutorial is here](https://spack-tutorial.readthedocs.io/en/latest/tutorial_developer_workflows.html)

Reply:
Thanks @alalazo, I've changed to `type='build'` for CMake and cleaned up the `type` for a couple of other dependencies too.